spec/src/main/asciidoc/microprofile-lra-spec.adoc

```diff
@@ -555,11 +555,143 @@ of these methods. If it knows it will never be able to compensate or complete th
+
+[[jaxrs-response-table]]
+|===
+| Annotation | HTTP method | Expected status codes | Response
+
+| `@Compensate`
+| PUT
+| 200, 202, 404, 412
+| <<source-ParticipantStatus,ParticipantStatus>>
+
+| `@Complete`
+| PUT
+| 200, 202, 404, 412
+| <<source-ParticipantStatus,ParticipantStatus>>
+
+| `@Status`
+| GET
+| 200, 202, 404, 412
+| <<source-ParticipantStatus,ParticipantStatus>>
+
+| `@Forget`
+| DELETE
+| 200, 404, 412
+| no expectations
+
+|=== 
+
+If the participant annotation is not accompanied by an associated JAX-RS HTTP method 
+annotation according to the <<jaxrs-response-table,table above>>, the error SHOULD be reported using a JAX-RS 
+exception mapper that maps to a `412 Precondition Failed` HTTP status code.
+
+[[non-jaxrs-participant-methods]]
+===== Non-JAX-RS participant methods
+
+When the participant annotations are applied to the non-JAX-RS resource methods they 
+MUST adhere to these predefined signatures:
+
+* *Return type*: 
+** `void`: successfull execution is mapped to `Compensated` or `Completed` participant statuses, 
+ error execution is handled by exception mapping
+*** applicable only for `@Compensate` and `@Complete` participant methods
+** <<source-ParticipantStatus,ParticipantStatus>>
+** `javax.ws.rs.core.Response`: handled similarly as for 
+<<jaxrs-participant-methods, JAX-RS participant methods>>
+** `java.util.concurrent.CompletionStage`: with the parameter of any of the previously 
+defined types
+* *Arguments*: up to 2 arguments of types in this order:
+** `java.net.URI`: representing current LRA context identification
+** `java.net.URI`: representing potentional parent LRA context identification
+
+Declaring more than two arguments, different types of arguments or different return type 
+for any non-JAX-RS method annotatated with the participant marker annotation MUST result 
+in the deployment time `InvalidLRAParticipantDefinitionException`. 
+Please note that both arguments are optional but the order is required.
+
+Examples of valid signatures:
+
+[source,java]
+----
+@Compensate
+public void compensate(URI lraId, URI parentId)
+
+@Complete
+public Response complete(URI lraId)
+
+@Status
+public CompletionStage<ParticipantStatus> status(URI lraId)
+----
+
+Examples of invalid signatures:
+
+[source,java]
+----
+@Compensate
+public void compensate(String lraId, String parentId) // invalid types of arguments
+
+@Compensate
+public String compensate(URI lraId) // invalid return type
+
+@Forget
+public void forget(URI lraId, URI parentId, String additional) // too many arguments
+----
+
+If any of the described methods throws an exception, we distinguish two cases depending 
+on the exception type:
+
+* `WebApplicationException` -- the exception is mapped to the HTTP response it carries 
+and then handled as defined in the section 
+<<jaxrs-participant-methods, JAX-RS participant methods>>
+* *any other exception* results into `FailedToComplete` or `FailedToCompensate` 
```

Comment:
in `@Forget` it does not matter how the method ends. But why for `@Status` cannot implementation return FailedTo* states?

Reply:
So if @Forget does not matter then why would one infer that the state is FailedToComplete if it throws an exception?
And throwing an exception from @Status just means that there was a problem obtaining the status - it says nothing about what the actual status is.

Reply:
so this mapping should be applicable only for complete and compensate and if any other exception then WebApplicationException is thrown in Status what should be returned by the proxy? And exceptions in Forget can be ignored, because proxy is not required to return anything to the coordination.

Reply:
I think it is valid to infer `FailedToCompensate` if the compensate method throws an unexpected exception. But if the forget or status methods throw exceptions then the participant may already have compensated or is in the process of compensating and telling the coordinator the wrong thing is either suppressing failure conditions or is reporting false positives, both of which are undesirable.

I do notice that the `@Status` and `@Forget` methods are under-specified but I would choose to report one of the HTTP 500 codes and leave it up to the implementation to decided what behaviour is preferable, but if it not retry then it SHOULD log the failure.

Reply:
The latest commit still has the same text about what happens if an exception is thrown from the status and forget methods:

> any other exception results into FailedToComplete or FailedToCompensate participant states